File: contest/LeetCode_Weekly_Contest_206/Check_If_String_Is_Transformable_With_Substring_Sort_Operations.py
class Solution:
    def isTransformable(self, s: str, t: str) -> bool:
        n = len(s)
        posi = [n] * 10
        for i, c1 in enumerate(s):
            v = ord(c1) - ord('0')
            if posi[v] == n:
                posi[v] = i
        for c2 in t:
            v = ord(c2) - ord('0')
            if posi[v] == n:
                return False
            for i in range(v):
                if posi[i] < posi[v]:
                    return False
            posi[v] += 1
            while posi[v] < n and s[posi[v]] != c2:
                posi[v] += 1
        return True

File: contest/LeetCode_Weekly_Contest_206/Check_If_String_Is_Transformable_With_Substring_Sort_Operations.cpp
class Solution {
public:
    vector<int> pos[10];
    int n;
    bool isTransformable(string s, string t) {
        n = s.length();
        
        for (int i = 0; i < n; ++i) {
            int idx = s[i] - '0';
            pos[idx].push_back(i);
        }
        
        vector<int> p(10, 0);
        
        for (int i = 0; i < t.length(); ++i) {
            int idx = t[i] - '0';
            if (p[idx] >= pos[idx].size()) return false;
            
            int where = pos[idx][p[idx]];
            p[idx]++;
            
            for (int v = 0; v < idx; ++v) {
                if (p[v] >= pos[v].size()) continue;
                if (pos[v][p[v]] < where) return false;
            }
        }
        
        return true;
    }
};
